add hirsute dev setup support, clean up ubuntu version identification, other small fixes

e-mcginnis opened the pull request:
- use focal packages for hirsute where required for clang-10 and
apache-arrow
- RELEASE and RELEASE_CODENAME were the same thing, only use
RELEASE_CODENAME
- pip would throw an error about testresources not being in the user
install directory, add it
- clangd-10 is required for lsp servers, add it to the dev tools

l-hoang commented on libquery/include/querying/ExpressionTree.h:
```diff
@@ -90,64 +140,114 @@ class ArithmeticTree {
+//! Since a compiler parse can possibly return different tree types, this
+//! typedef exists. Note it is typed to CompilerTreePossibilityIndex, so make
+//! sure that it stays consistent. (e.g., be careful with changing the order)
+using CompilerTreePossibility = std::variant<std::unique_ptr<ExpressionTree>,
```
agreed

You may have noticed by now but my naming sensibilities aren't the greatest.

Will change.

l-hoang commented on libquery/src/CypherCompiler.cpp:
```diff
@@ -91,62 +91,78 @@ QueryProperty ParseQueryProperty(CypherASTNode expression_ast) {
-    } else if (arg_number == 1) {
-      // TODO(l-hoang)
-    }
+    GALOIS_LOG_WARN("Parsed {} as the int", int_data);
```
LOG has 5 return levels.

Debug
Verbose
Info (can't be used because disabled)
Warning
Error

I would have liked to use Info, but it's not being used anywhere. Now that you mention it Verbose probably fits this way better. I wonder why I chose to use Warning. It was probably because I thought Verbose was too low of a level?

Will change.

l-hoang commented on libquery/src/CypherCompiler.cpp:
```diff
@@ -91,62 +91,78 @@ QueryProperty ParseQueryProperty(CypherASTNode expression_ast) {
+    // if in debug build just go ahead and fail
+    assert(false);
+    // Probably should not get here (hence the "undefined behavior" comment)
+    return std::unique_ptr<ArithmeticTree>();
```
I wasn't sure if I wanted to potentially recover from this error down the line; let me check the context again.

There are many instances of this in the code where FATAL would be appropriate.

l-hoang replied:
In this case, yes I would support FATAL and have changed it.

e-mcginnis committed:
add hirsute dev setup support, clean up ubuntu version identification,
other small fixes

- use focal packages for hirsute where required for clang-10 and
apache-arrow
- RELEASE and RELEASE_CODENAME were the same thing, only use
RELEASE_CODENAME
- pip would throw an error about testresources not being in the user
install directory, add it

arthurp commented on scripts/setup_ubuntu.sh:
```diff
@@ -49,25 +50,39 @@ run_as_original_user() {
-VERSION=$(lsb_release --release --short | cut -d . -f 1)
-if [ "$VERSION" == "16" ]; then
+# Clang-10 isn't present in the hirsute repos any more, so we need to add the upstream repo
+if [[ "$RELEASE_CODENAME" == "xenial" ||  "$RELEASE_CODENAME" == "hirsute" ]]; then
```
I would prefer to use numbers. It's much easier to keep track of.

arthurp commented on scripts/setup_ubuntu.sh:
```diff
@@ -49,25 +50,39 @@ run_as_original_user() {
-curl "https://apache.jfrog.io/artifactory/arrow/$RELEASE_ID/apache-arrow-apt-source-latest-$RELEASE_CODENAME.deb" \
-  --output /tmp/apache-arrow-apt-source-latest.deb
+
+if [ "$RELEASE_CODENAME" == "hirsute" ]; then
```
numbers please.

arthurp commented on scripts/setup_ubuntu.sh:
```diff
@@ -91,12 +106,15 @@ else
+run_as_original_user pip3 install testresources conan==1.36 PyGithub packaging
+# since we install as $USER, pip installs ~/.local/bin,
+# but this is not in the default ubuntu $PATH. Add it.
+run_as_original_user export PATH=$PATH:$HOME/.local/bin
```
I think you are trying to add it to the user path. But this will not actually do anything (because of how shells work). Also I'm pretty sure `.local/bin` is in the user path by default at least if the directory exists.

e-mcginnis replied:
bah, you are definitely right about export. 
re: $PATH, when I ran these scripts on my 21.04 box, the next conan script failed because it could not find `conan` despite it being available in `~/.local/bin`. 
I suspect if I had opened a fresh shell, whatever script adds `~/.local/bin` to $PATH automagically would have done its job. 
anyway, I'll drop this change for now

arthurp commented on scripts/setup_ubuntu.sh:
```diff
@@ -116,7 +134,7 @@ apt install -yq --allow-downgrades \
   apt install -yq gcc-9 g++-9 clang-10
   # in newer versions of ubuntu clang++-10 is installed as a part of clang-10
-  if [ "$VERSION" == "16" ]; then
+  if [[ "$RELEASE_CODENAME" == "xenial" ]]; then
```
numbers.

arthurp reviewed (changes requested):
We discussed numbers vs names when we wrote these scripts and we chose numbers because it's easier to remember which version it is by number because they are just years. So please don't change it to names.

e-mcginnis commented:
hm, didn't mean to close it...

arthurp commented on scripts/setup_ubuntu.sh:
```diff
@@ -49,25 +51,39 @@ run_as_original_user() {
+    curl "https://apache.jfrog.io/artifactory/arrow/$RELEASE_ID/apache-arrow-apt-source-latest-focal.deb" \
+         --output /tmp/apache-arrow-apt-source-latest.deb
+    wget "mirrors.kernel.org/ubuntu/pool/main/r/re2/libre2-5_20200101+dfsg-1build1_amd64.deb" \
+         -O /tmp/libre2-5.deb
```
Use curl

arthurp replied:
Also don't use a mirror. Use the main ubuntu repo for this.

e-mcginnis replied:
funny enough, I actually used wget because 
```
curl "http://mirrors.kernel.org/ubuntu/pool/main/r/re2/libre2-5_20200101+dfsg-1build1_amd64.deb" \
         --output /tmp/libre2-5.deb
```
results in a 0 byte file, whereas `wget` just worked

that said, changing to the main ubuntu repo 
```
curl "http://archive.ubuntu.com/ubuntu/pool/main/r/re2/libre2-5_20200101+dfsg-1build1_amd64.deb" --output libre2-5.deb```
fixes that issue, so I guess we get two birds with one stone.

arthurp commented on scripts/setup_ubuntu.sh:
```diff
@@ -49,25 +51,39 @@ run_as_original_user() {
+         --output /tmp/apache-arrow-apt-source-latest.deb
+    wget "mirrors.kernel.org/ubuntu/pool/main/r/re2/libre2-5_20200101+dfsg-1build1_amd64.deb" \
+         -O /tmp/libre2-5.deb
+    apt install -yq /tmp/libre2-5.deb && rm /tmp/libre2-5.deb
```
There is a way to add an older releases repo to apt and then access it using `libre2/focal` as the package descriptor. I would prefer that.

e-mcginnis replied:
something like that can definitely be done with apt .pref files:
1) add the focal apt repo source.list
2) add the focal.pref file, something like:
```
# only allow this package if no other version is installed
Package: libre2-5
Pin: release n=focal
Pin-Priority: 50

# never use other packages from focal
Package: *
Pin: release n=focal
Pin-Priority: -1
```

That said, I'm not convinced that adding the full focal repo and pref file actually results in a cleaner solution vs just grabbing the one package we need and installing it.
We really only need libre2-5 until apache arrow releases a 21.04 package, which will likely depend on libre2-9 instead.
Just grabbing the single package we need makes the future clean up straight forward, just remove libre2-5 (it might even be added to the autoremove list when we install the newer apache arrow, but I'm not 100% confident on that). 

Of course, I'm open to discussion here.

arthurp replied:
Fair. Can you put a comment here if you are making other changes? Because if we start using more than one file from other releases it would be worth setting up a nice way to pull them in.

BTW, I think you are over complicating things. I think you can reasonably safely just add the source line for the old release and then pull in packages using the special version syntax I mentioned. The trick you are showing is only needed if you want to selectively get packages from a NEWER release since those will generally have newer version numbers and hense override the older release by default.

e-mcginnis replied:
I tested this out on my hirsute install. Added the repo with `sudo apt-add-repository "deb http://us.archive.ubuntu.com/ubuntu/ focal main restricted"` which came right from my focal machines sources.list


apt shows we are getting `libre2-5` from focal:
```
apt list libre2-5
Listing... Done
libre2-5/focal 20200101+dfsg-1build1 amd64
libre2-5/focal 20200101+dfsg-1build1 i386
```
but since the package is only available from focal, we don't have to specify the `libre2-5/focal` and apt assumes we just want the only available package instead. 
```
❯ sudo apt install libre2-5
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  libre2-5
0 upgraded, 1 newly installed, 0 to remove and 15 not upgraded.
Need to get 162 kB of archives.
After this operation, 485 kB of additional disk space will be used.
Get:1 http://us.archive.ubuntu.com/ubuntu focal/main amd64 libre2-5 amd64 20200101+dfsg-1build1 [162 kB]
Fetched 162 kB in 0s (476 kB/s)
Selecting previously unselected package libre2-5:amd64.
.....
```
The side effect of this being now any package no longer available in hirsute will get pulled automatically from focal, instead of making the user aware the package is depreciated for their system. This definitely can lead to worse FrankenDebian without the user knowing. I would prefer to require an active step before pulling in old releases packages, which would require the `.pref` file.

arthurp replied:
Valid point. I didn't think about how the package wouldn't exist at all in the newer verion.

The direct download approach is probably best for now.

We eventually need to move away from using all these external repos for dependencies of our packages. Though if some stay around for the build environment it's not a big deal.

arthurp reviewed (approved):
LGTM.
